src/app/helpers/formTemplate/formTemplate.ts

```diff
@@ -0,0 +1,118 @@
+import { OnInit } from "@angular/core";
+import { ActivatedRoute, Router } from "@angular/router";
+import { ToastrService } from "ngx-toastr";
+import { Observable } from "rxjs";
+import { takeUntil } from "rxjs/operators";
+import { WithFormCheck } from "src/app/guards/form/form.guard";
+import { AbstractModel } from "src/app/models/AbstractModel";
+import { ApiErrorDetails } from "src/app/services/baw-api/api.interceptor.service";
+import { ResolvedModel } from "src/app/services/baw-api/resolver-common";
+import { PageComponent } from "../page/pageComponent";
+
+export abstract class FormTemplate<M extends AbstractModel, F>
+  extends WithFormCheck(PageComponent)
+  implements OnInit {
+  public loading: boolean;
+  public models: { [key: string]: AbstractModel };
+  public schema: FormlySchema;
+
+  /**
+   * @param modelKeys Model resolver keys
+   * @param api API Service
+   * @param notifications ToastrService
+   * @param route ActivatedRoute
+   * @param router Router
+   */
+  constructor(
+    protected modelKeys: string[],
+    protected fields: any[],
+    protected notifications: ToastrService,
+    protected route: ActivatedRoute,
+    protected router: Router
+  ) {
+    super();
+  }
+
+  ngOnInit() {
+    // Grab models from route data
+    const data = this.route.snapshot.data;
+    const resolvedModels: ResolvedModel<AbstractModel>[] = this.modelKeys.map(
```

Comment:
does data ever contain non-models?

If it doesn't, then there's no need to specify keys?

Reply:
Data does contain non-models (specifically it contains all of the page data as well), but ill move to your suggestion because the page data shouldn't affect the behaviour